src/modm/driver/adc/ads816x_impl.hpp

```diff
@@ -0,0 +1,134 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2021, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ADS816X_HPP
+#	error "Don't include this file directly! Use 'ads816x.hpp' instead."
+#endif
+
+#include <bit>
+
+namespace modm
+{
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::initialize(Mode mode)
+{
+	using namespace std::chrono_literals;
+	RF_BEGIN();
+
+	// Unlock register access
+	RF_CALL(registerAccess(Command::Write, Register::REG_ACCESS, 0b1010'1010));
+
+	timeout.restart(4us);
+	RF_WAIT_UNTIL(timeout.isExpired());
+
+	// Set mode
+	RF_CALL(registerAccess(Command::Write, Register::DEVICE_CFG, uint8_t(mode)));
+
+	RF_END();
+}
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<uint16_t>
+Ads816x<SpiMaster, Cs>::manualModeConversion(uint8_t nextChannel)
+{
+	RF_BEGIN();
+
+	RF_WAIT_UNTIL(this->acquireMaster());
+	Cs::reset();
+
+	buffer[0] = uint8_t(Command::Write) | (uint16_t(Register::CHANNEL_ID) >> 8);
+	buffer[1] = uint16_t(Register::CHANNEL_ID);
+	buffer[2] = (nextChannel & 0b0000'0111);
+
+	RF_CALL(SpiMaster::transfer(buffer, buffer2, 3));
+
+	if (this->releaseMaster()) {
+		Cs::set();
+	}
+
+	RF_END_RETURN( static_cast<uint16_t>((static_cast<uint16_t>(buffer2[0]) << 8) | buffer2[1]) );
+}
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::autoSequenceConversion(uint8_t channelsBitmask, std::array<uint16_t, 8>& result)
+{
+	using namespace std::chrono_literals;
+	RF_BEGIN();
+
+	// Write channel bitmask config (AUTO_SEQ_CH)
+	RF_CALL(registerAccess(Command::Write, Register::AUTO_SEQ_CFG1, channelsBitmask));
+
+	timeout.restart(4us);
+	RF_WAIT_UNTIL(timeout.isExpired());
+
+	// Set SEQ_START (0b1 << 0) bit
+	RF_CALL(registerAccess(Command::SetBits, Register::SEQ_START, 0b1));
+
+	// Read conversion results
+	for (buffer2[0] = 0; buffer2[0] < std::popcount(channelsBitmask); buffer2[0]++) {
+		timeout.restart(4us);
+		RF_WAIT_UNTIL(timeout.isExpired());
+
+		RF_WAIT_UNTIL(this->acquireMaster());
+		Cs::reset();
+
+		RF_CALL(SpiMaster::transfer(nullptr, buffer, 2));
```

Comment:
If I am reading the datasheet correctly, the first run of the loop will not read the correct sample. According to the sequence diagram the cycle after setting `SEQ_START` will not output the first sample from the sequence but from the input configured before:
![image](https://user-images.githubusercontent.com/25187160/147417262-d771a660-8f56-4c51-9ee5-639aea444106.png)

Data from the first sequence channel is output on the fourth transfer, not the third.

Reply:
Correct. I already fixed that behavior in https://github.com/rleh/modm/commit/0eaa41dc82d945e783cbd1d6e073c0533c5b76a8.

But in _auto sequence_ mode the chip still seems to return measurements of the first channel...